### PR TITLE
[Feature] Launch Long Live Servers and Multiple Client Groups

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

Features in this PR:
1. enable launch long live servers and multiple client groups simultaneously. Sequentially launch is also supported.
2. get available port instead of fixed which could be unavailable sometimes.
3. comment out one line in `train_dist.py` which could cause error when running. Discussed with JJ, he will fix it in another PR.

[How to Launch]
For end users, in order to launch long live server, just specify `--keep_alive --server_name xxx` besides normal command.  that's all. If user want to launch other groups which utilize exiting alive servers, just specify `--server_name xxx` which is the same name as first launch. that's all.
Please note, the first launch process/shell will be alive until user send `ctrl-c` interrupt signal to it. then servers will be terminated by existing clean-up logic. This should be refined in the near future: shut down servers more elegantly. And current clean-up logic has a side-effect: any other launch crashes or interrupted, all processes including the long live servers will be killed in force.

[Examples]
launch long live server(first group):
`DGL_HOME=/home/ubuntu/workspace/dgl_p1 DGL_LIBRARY_PATH=$DGL_HOME/build PYTHONPATH=$DGL_HOME/python:$PYTHONPATH DGLBACKEND=pytorch python3 ~/workspace/dgl_p1/tools/launch.py --workspace ~/workspace/dgl_p1/examples/pytorch/graphsage/experimental/ --keep_alive --server_name ruying_test --num_trainers 1 --num_samplers 0 --num_servers 1 --part_config data/ogb-product.json --ip_config ~/workspace/ip_config.txt "python3 train_dist.py --graph_name ogb-product --ip_config ~/workspace/ip_config.txt --num_epochs 5 --batch_size 1000"`
launch another group(2nd one):
`DGL_HOME=/home/ubuntu/workspace/dgl_p1 DGL_LIBRARY_PATH=$DGL_HOME/build PYTHONPATH=$DGL_HOME/python:$PYTHONPATH DGLBACKEND=pytorch python3 ~/workspace/dgl_p1/tools/launch.py --workspace ~/workspace/dgl_p1/examples/pytorch/graphsage/experimental/ --server_name ruying_test --num_trainers 1 --num_samplers 0 --num_servers 1 --part_config data/ogb-product.json --ip_config ~/workspace/ip_config.txt "python3 train_dist.py --graph_name ogb-product --ip_config ~/workspace/ip_config.txt --num_epochs 5 --batch_size 1000"`

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/graphsage/experimental/train_dist.py [hunk 1]
```diff
@@ -184,7 +184,8 @@ def pad_data(nids, device):
 def run(args, device, data):
     # Unpack data
     train_nid, val_nid, test_nid, in_feats, n_classes, g = data
-    train_nid = pad_data(train_nid, device)
+    # Below line causes error sometimes.
+    # train_nid = pad_data(train_nid, device)
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist.py:188]: what does it do?

File: python/dgl/distributed/dist_context.py [hunk 2]
```diff
@@ -228,7 +228,7 @@ def initialize(ip_config, num_servers=1, num_workers=0,
         formats = os.environ.get('DGL_GRAPH_FORMAT', 'csc').split(',')
         formats = [f.strip() for f in formats]
         rpc.reset()
-        keep_alive = os.environ.get('DGL_KEEP_ALIVE') is not None
+        keep_alive = bool(int(os.environ.get('DGL_KEEP_ALIVE', 0)))
```
Comment (zheng-da) [inline python/dgl/distributed/dist_context.py:231]: why do you need to convert it to int and then to bool?

File: tools/launch.py [hunk 3]
```diff
@@ -424,6 +432,59 @@ def wrap_cmd_with_extra_envvars(cmd: str, env_vars: list) -> str:
     env_vars = " ".join(env_vars)
     return wrap_cmd_with_local_envvars(cmd, env_vars)
 
+
+g_monitor_file = None
+g_group_id = 0
+
+def has_alive_servers(args):
+    """Check whether there exist alive servers and get available group ID if exist."""
```
Comment (zheng-da) [inline tools/launch.py:440]: can you say more about how you check it?

File: tools/launch.py [hunk 4]
```diff
@@ -457,23 +518,27 @@ def submit_jobs(args, udf_command):
...
+        )
+        for i in range(len(hosts) * server_count_per_machine):
+            ip, _ = hosts[int(i / server_count_per_machine)]
+            server_env_vars_cur = f"{server_env_vars} DGL_SERVER_ID={i}"
+            cmd = wrap_cmd_with_local_envvars(udf_command, server_env_vars_cur)
+            cmd = wrap_cmd_with_extra_envvars(cmd, args.extra_envs) if len(args.extra_envs) > 0 else cmd
+            cmd = 'cd ' + str(args.workspace) + '; ' + cmd
+            thread_list.append(execute_remote(cmd, ip, args.ssh_port, username=args.ssh_username))
+    else:
+        print("Alive servers whose name is {} are found, let's use them direclty.".format(args.server_name))
```
Comment (zheng-da) [inline tools/launch.py:541]: how do you know if the servers have the graph data the trainers want?

File: examples/pytorch/graphsage/experimental/train_dist.py (hunk 1 above)
Comment (VoVAllen) [reply examples/pytorch/graphsage/experimental/train_dist.py:188]: This is a fix when batch size is small. The nid might be unequal between trainers, and when batch size is small, the number of batches will be unequal across trainer, which results in hanging for the trainer with more batches. 
Here I pad the nid to the same length (as the max one across trainer), by repeating the starting ids. But here cannot handle duplicate node ids within the same batches, thus might cause error when shuffling.

Now we decided to adopt drop last strategy, that drop the batch for the trainer with more batches than others

Comment (VoVAllen) [reply examples/pytorch/graphsage/experimental/train_dist.py:188]: This is a fix when batch size is small. The nid might be unequal between trainers, and when batch size is small, the number of batches will be unequal across trainer, which results in hanging for the trainer with more batches. 
Here I pad the nid to the same length (as the max one across trainer), by repeating the starting ids. But here cannot handle duplicate node ids within the same batches, thus might cause error when shuffling.

Now we decided to adopt drop last strategy, that drop the batch for the trainer with more batches than others

Comment (VoVAllen) [reply examples/pytorch/graphsage/experimental/train_dist.py:188]: Issue: https://github.com/dmlc/dgl/issues/3538

File: python/dgl/distributed/dist_context.py (hunk 2 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/dist_context.py:231]: because `int` value is specified for `DGL_KEEP_ALIVE` and `os.environ.get()` always return string? so has to convert to int first. Then check whether the int value equals zero. it's a bit clumsy.

File: tools/launch.py (hunk 3 above)
Comment (Rhett-Ying) [reply tools/launch.py:440]: sure.

File: tools/launch.py (hunk 4 above)
Comment (Rhett-Ying) [reply tools/launch.py:541]: if long live servers are utilized, the user should make sure the graph data they want has already exist when the long live servers are booted. Namely, all the trainers share the DistGraphServer and the graph data in it.

File: python/dgl/distributed/dist_context.py (hunk 2 above)
Comment (jermainewang) [reply python/dgl/distributed/dist_context.py:231]: How about: `keep_alive = bool(os.environ.get('DGL_KEEP_ALIVE', False))`

Comment (Rhett-Ying) [reply python/dgl/distributed/dist_context.py:231]: if so, `DGL_KEEP_ALIVE=False` will be evaluated as `True`. namely, once macro `DGL_KEEP_ALIVE` is defined, it's always evaluated as `True`.

File: tools/launch.py [hunk 5]
```diff
@@ -424,6 +432,68 @@ def wrap_cmd_with_extra_envvars(cmd: str, env_vars: list) -> str:
...
+    launch new servers or utilize the existing alive ones. If there
+    exist alive servers, we obtain availale group ID from the monitor
+    file which could be used in current client groups.
+
+    """
+    if args.server_name is None:
+        return False
+    global g_monitor_file
+    global g_group_id
+    monitor_file = '/tmp/dgl_dist_monitor_' + args.server_name
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply tools/launch.py:454]: Consider using https://docs.python.org/3/library/tempfile.html for temporary file.

File: tools/launch.py [hunk 6]
```diff
@@ -424,6 +432,68 @@ def wrap_cmd_with_extra_envvars(cmd: str, env_vars: list) -> str:
...
+    try:
+        if g_monitor_file is not None:
+            os.remove(g_monitor_file)
+            os.remove(g_monitor_file + '.lock')
+            print("Monitor file for alive servers is removed: {}.".format(g_monitor_file))
+    except:
+        print("Failed to delete monitor file for alive servers: {}.".format(g_monitor_file))
+
+def get_available_port(ip):
+    """Get available port with specified ip."""
```
Comment (jermainewang) [inline tools/launch.py:487]: This post may be useful for finding an available port: https://stackoverflow.com/questions/1365265/on-localhost-how-do-i-pick-a-free-port-number

File: tools/launch.py (hunk 4 above)
Comment (jermainewang) [reply tools/launch.py:550]:
```suggestion
        print(f"Use running server {args.server_name}."
```

**Review by jermainewang — APPROVED** [review]
Overall LGTM. Some minor suggestions. I will do another review once the corresponding user guide chapter is updated.

File: tools/launch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/launch.py:454]: windows is not supported at all in DistDGL.
A fixed name/path is required to be checked for each launch while `tempfile` is always random and unique for each launch?

File: tools/launch.py [hunk 7]
```diff
@@ -424,6 +432,68 @@ def wrap_cmd_with_extra_envvars(cmd: str, env_vars: list) -> str:
...
+def has_alive_servers(args):
+    """Check whether there exist alive servers and get available group ID if exist.
+
+    For each group of long live servers, a monitor file named
+    'dgl_dist_monitor_{args.server_name}' is created under '/tmp/' directory.
+    We check the existence of this monitor file to determine whether to
+    launch new servers or utilize the existing alive ones. If there
+    exist alive servers, we obtain availale group ID from the monitor
+    file which could be used in current client groups.
+
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/launch.py:448]: fixed.

File: tools/launch.py [hunk 8]
```diff
@@ -424,6 +432,68 @@ def wrap_cmd_with_extra_envvars(cmd: str, env_vars: list) -> str:
     env_vars = " ".join(env_vars)
     return wrap_cmd_with_local_envvars(cmd, env_vars)
 
+
+g_monitor_file = None
+g_group_id = 0
+
+def has_alive_servers(args):
+    """Check whether there exist alive servers and get available group ID if exist.
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/launch.py:440]: fixed.

File: tools/launch.py (hunk 6 above)
Comment (Rhett-Ying) [reply tools/launch.py:487]: the post you mentioned is helpful for getting available port on current host while we need available port for a specified host. we may have multiple hosts.

File: python/dgl/distributed/dist_context.py (hunk 2 above)
Comment (jermainewang) [reply python/dgl/distributed/dist_context.py:231]: Use `ast.literal_eval`?

File: tools/launch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply tools/launch.py:458]: I see. Then tempfile is not suitable for your case.

File: docs/source/guide/distributed-tools.rst [hunk 9]
```diff
@@ -75,3 +75,44 @@ The launch script creates a specified number of training jobs (``--num_trainers`
 In addition, a user needs to specify the number of sampler processes for each trainer
 (``--num_samplers``). The number of sampler processes has to match with the number of worker processes
 specified in :func:`~dgl.distributed.initialize`.
+
+Below shows an example of launching long live servers and multiple distributed training jobs
+in a cluster. ``Long Live Servers`` means :class:`~dgl.distributed.DistGraph` servers are launched
+once only and kept alive even training jobs finish and exit. Then any following training jobs
+could fetch graph structures and data from the same servers. As a result, we do not need to load
+graph multiple times which may take a lot of time for large graph.
```
Comment (jermainewang) [inline docs/source/guide/distributed-tools.rst:83]:
```
It is common that users may want to try different models or training configurations against the same graph data. 
To avoid repetitively loading the same graph data, DGL allows users to launch a persistent graph server to be 
shared across multiple training jobs. A persistent graph server will stay alive even all training workers have 
finished and exited. Below shows an example of launching a persistent graph server:
```

File: docs/source/guide/distributed-tools.rst [hunk 10]
```diff
@@ -75,3 +75,44 @@ The launch script creates a specified number of training jobs (``--num_trainers`
...
+
+Launch long live servers and first group of distributed training job.
```
Comment (jermainewang) [inline docs/source/guide/distributed-tools.rst:85]:
```
We first launch the graph server together with the first group of training workers.
```

File: docs/source/guide/distributed-tools.rst [hunk 11]
```diff
@@ -75,3 +75,44 @@ The launch script creates a specified number of training jobs (``--num_trainers`
...
+    --num_samplers 4 \
+    --num_servers 1 \
+    --part_config data/ogb-product.json \
+    --ip_config ip_config.txt \
+    --keep_alive \
+    --server_name long_live \
+    "python3 code/train_dist.py --graph-name ogb-product --ip_config ip_config.txt --num-epochs 5 --batch-size 1000 --lr 0.1 --num_workers 4"
+
+``--keep_alive`` is specified explicitly which indidcates boot long live servers. ``--server_name``
+is the name of long live servers which could be used in other launches.
```
Comment (jermainewang) [inline docs/source/guide/distributed-tools.rst:101]:
```
Pay attention to the ``--keep_alive`` option, which indicates the server should stay alive after workers have
finished. ``--server_name`` is the given name of the server which will be referred when launching new training
jobs.
```

File: docs/source/guide/distributed-tools.rst [hunk 12]
```diff
@@ -75,3 +75,44 @@ The launch script creates a specified number of training jobs (``--num_trainers`
...
+
+Launch another group of distributed training job which utilize existing long live servers.
```
Comment (jermainewang) [inline docs/source/guide/distributed-tools.rst:103]:
```
Launch another group of distributed training job and connect to the existing persistent server.
```

File: docs/source/guide/distributed-tools.rst [hunk 13]
```diff
@@ -75,3 +75,44 @@ The launch script creates a specified number of training jobs (``--num_trainers`
...
+    --num_trainers 2 \
+    --num_samplers 4 \
+    --num_servers 1 \
+    --part_config data/ogb-product.json \
+    --ip_config ip_config.txt \
+    --server_name long_live \
+    "python3 code/train_dist.py --graph-name ogb-product --ip_config ip_config.txt --num-epochs 5 --batch-size 1000 --lr 0.1 --num_workers 4"
+
+All the arguments that are related to servers should be kept same as previous launch. But training
+specific arguments such as ``--num-epochs``, ``--batch-size`` and so on are free to change.
```
Comment (jermainewang) [inline docs/source/guide/distributed-tools.rst:118]: Can we let users omit those server arguments?

File: python/dgl/distributed/dist_context.py (hunk 2 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/dist_context.py:231]: return type of `literal_eval` is not very  clear. I prefer to convert on our own though Current parse is a bit weird.

File: docs/source/guide/distributed-tools.rst (hunk 9 above)
Comment (Rhett-Ying) [reply docs/source/guide/distributed-tools.rst:83]: accepted.

File: docs/source/guide/distributed-tools.rst (hunk 10 above)
Comment (Rhett-Ying) [reply docs/source/guide/distributed-tools.rst:85]: accepted.

File: docs/source/guide/distributed-tools.rst (hunk 11 above)
Comment (Rhett-Ying) [reply docs/source/guide/distributed-tools.rst:101]: accepted.

File: docs/source/guide/distributed-tools.rst (hunk 12 above)
Comment (Rhett-Ying) [reply docs/source/guide/distributed-tools.rst:103]: accepted.

**Review by jermainewang — APPROVED** [review]
